### PR TITLE
Use YAMLToJSONStrict instead of YAMLToJSON when loading patches

### PR DESCRIPTION
**Pull request by tom1299**
Currently when yamls with duplicate key entries are used for config patches, neither a warning nor an error message is displayed. For example using this malformed patch (See duplicate key in the last line) :
```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
kubeadmConfigPatches:
- |
  kind: InitConfiguration
  nodeRegistration:
    kubeletExtraArgs:
      anonymous-auth: "true"
      anonymous-auth: "false"
``` 
with `kind create cluster --config ./kind-config.yaml -v=9` will result in the following `kubeadm.config` (only relavant lines are shown):
```bash
  kubeletExtraArgs:
    anonymous-auth: "false"
``` 
This PR contains a change to use `YAMLToJSONStrict` instead of the more relaxed `YAMLToJSON`. I also added a test case for an example with a duplicate key.

### CODE REVIEW

**Comment by stmcginnis** [comment]
Some may expect the more lenient handling at this point. Not sure how big of an impact this would be.

**Comment by tom1299** [comment]
@stmcginnis I was thinking that as well. `kubeadm init` displays a warning but continues to setup the node when the `kubeadm.config` contains duplicate keys. I could alter the logic to first try strict, and if there was an error, display that and do it the lenient way. That would reflect the behavior of `kubeadm`.

**Comment by tom1299** [comment]
@stmcginnis I changed the implementation to a more permissive approach which just writes a warning to stderr:
```bash
$ kind create cluster --config ./kind-config.yaml 
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.34.0) 🖼
 ✓ Preparing nodes 📦  
⠈⠑ Writing configuration 📜 
WARN: Error converting patch of kind InitConfiguration to json: "yaml: unmarshal errors:
  line 5: key "anonymous-auth" already set in map". Trying more permissive conversion.
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
Set kubectl context to "kind-kind"
You can now use your cluster with:

kubectl cluster-info --context kind-kind

Thanks for using kind! 😊
```
The warning messes up the output a little bit though. Maybe there is a better way to log this error (?) This change would not impact existing malformed patches and would reflect the current behaviour of `kubeadm init` as this command also logs a warning and continues.

File: pkg/internal/patch/mergepatch.go [hunk 1]
```diff
@@ -44,9 +46,19 @@ func parseMergePatches(rawPatches []string) ([]mergePatch, error) {
 		if err != nil {
 			return nil, errors.WithStack(err)
 		}
-		json, err := yaml.YAMLToJSON([]byte(raw))
+		json, err := yaml.YAMLToJSONStrict([]byte(raw))
 		if err != nil {
-			return nil, errors.WithStack(err)
+			fmt.Fprintf(
```
Comment (stmcginnis) [inline pkg/internal/patch/mergepatch.go]: I do like this approach overall, but a little concerned about writing to stderr. Especially using format encoding that might not be supported on the terminal or output device where that is being sent to.

It does muddy the output to have warnings inline with progress information, but I think that part is fine. Could you switch to using `sigs.k8s.io/kind/pkg/log` and make this a `logger.Warn` call instead so output handling remains somewhat centralized?

Comment (BenTheElder) [reply pkg/internal/patch/mergepatch.go]: the logging utils will handle the output coordination

Comment (tom1299) [reply pkg/internal/patch/mergepatch.go]: Thanks for the advise. I changed the code to use the log module and also adjusted the log message to sound less alarming:
```go
logger.Warnf("Failed to strictly convert patch of kind %s to json: \"%v\". "+
	"Trying more permissive conversion.",
	matchInfo.Kind,
	err,
)
```
And the log looks much better now:
```bash
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.34.0) 🖼
 ✓ Preparing nodes 📦  
Failed to strictly convert patch of kind InitConfiguration to json: "yaml: unmarshal errors:
  line 5: key "anonymous-auth" already set in map". Trying more permissive conversion.
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
Set kubectl context to "kind-kind"
```
Although I had to add an additional parameter to some functions to pass the logger to the mergepatch.

File: pkg/internal/patch/kubeyaml.go [hunk 2]
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 package patch
 
 import (
+	"sigs.k8s.io/kind/pkg/log"
```
Comment (stmcginnis) [inline pkg/internal/patch/kubeyaml.go]: I believe this will cause linting failures. Need to have the separate stdlib, third party lib, local grouping.

File: pkg/internal/patch/mergepatch.go [hunk 3]
```diff
@@ -17,9 +17,9 @@ limitations under the License.
 package patch
 
 import (
-	"sigs.k8s.io/yaml"
-
 	"sigs.k8s.io/kind/pkg/errors"
+	"sigs.k8s.io/kind/pkg/log"
+	"sigs.k8s.io/yaml"
```
Comment (stmcginnis) [inline pkg/internal/patch/mergepatch.go]: This too.

**Review by stmcginnis — COMMENTED** [review]
A couple small things that will cause CI failure.

Not sure if we have a better way to get the logger instance. I know some people using `kind` as a library, so I don't love changing the method signatures. But if that's what we have to do...

File: pkg/internal/patch/mergepatch.go (hunk 3 above)
Comment (tom1299) [reply pkg/internal/patch/mergepatch.go]: Thanks for spotting that. I corrected the imports

**Comment by tom1299** [comment]
I didn't like changing the signatures of the functions either. Especially for such a small change. But I couldn't really find a better way.

**Comment by stmcginnis** [comment]
/ok-to-test

@BenTheElder @aojea what do you think about the method signature change? I think it may be worth it to have this output. I actually have run into an instance where some templating errors were duplicating keys and it was not at all obvious why the resulting cluster wasn't configured right.

**Comment by aojea** [comment]
+1 to strict parsing, I got bitted several times because of that and it takes a lot of time to figure out the problem

@BenTheElder is the best person for the plumbing of the logger

**Comment by BenTheElder** [comment]
> +1 to strict parsing, I got bitted several times because of that and it takes a lot of time to figure out the problem

Yeah, I'm generally a fan of strict parse but ...  it has bitten during enablement in the past.

I think we should hold off temporarily since we have some other critical fixes that we wouldn't want to cause adoption issues for but I think this makes sense long term
